Makefile

```diff
@@ -1,199 +1,176 @@
+  MD5_CODE := 
+  MD5_DATA := 
+  OPTIONS := 
+  EXTRA_DEPS :=
+endef
+
+define ucode_rule
+  # Variables defined outside the function need one dollar sign, whereas
+  # variables defined within the function need two. This is because make first
+  # expands all this text, substituting single-dollar-sign variables, and then
+  # executes all of it, causing all the assignments to actually happen.
+  ifeq ($(NAME),)
+   $$(error Microcode name not set!)
+  endif
+  UCODE_OUTPUT_DIR := $(PARENT_OUTPUT_DIR)/$(NAME)
+  ifeq ($(OS),Windows_NT)
+   UCODE_OUTPUT_DIR := $$(subst /,\,$$(UCODE_OUTPUT_DIR))
+  endif
+  CODE_FILE := $$(UCODE_OUTPUT_DIR)/$(NAME).code
+  DATA_FILE := $$(UCODE_OUTPUT_DIR)/$(NAME).data
+  SYM_FILE  := $$(UCODE_OUTPUT_DIR)/$(NAME).sym
+  TEMP_FILE := $$(UCODE_OUTPUT_DIR)/$(NAME).tmp.s
+  ALL_UCODES += $(NAME)
+  ifneq ($(MD5_CODE),)
+   ALL_UCODES_WITH_MD5S += $(NAME)
+  endif
+  ALL_OUTPUT_DIRS += $$(UCODE_OUTPUT_DIR)
+  OFF_OPTIONS := $(filter-out $(OPTIONS),$(ALL_OPTIONS))
+  OPTIONS_EQU := 
+  $$(foreach option,$(OPTIONS),$$(eval OPTIONS_EQU += -equ $$(option) 1))
+  OFF_OPTIONS_EQU := 
+  $$(foreach o2,$$(OFF_OPTIONS),$$(eval OFF_OPTIONS_EQU += -equ $$(o2) 0))
+  ARMIPS_CMDLINE := \
+   -strequ CODE_FILE $$(CODE_FILE) \
+   -strequ DATA_FILE $$(DATA_FILE) \
+   $$(OPTIONS_EQU) \
+   $$(OFF_OPTIONS_EQU) \
+   f3dex2.s \
+   -sym2 $$(SYM_FILE) \
+   -temp $$(TEMP_FILE)
+  # Microcode target
+  .PHONY: $(NAME)
+  $(NAME): $$(CODE_FILE)
+  # Directory target variables, see below.
+  $$(UCODE_OUTPUT_DIR): UCODE_OUTPUT_DIR:=$$(UCODE_OUTPUT_DIR)
+  # Directory target recipe
+  $$(UCODE_OUTPUT_DIR):
+	@echo "$(INFO)Creating directory $$(UCODE_OUTPUT_DIR)$(NO_COL)"
+  ifeq ($(OS),Windows_NT)
+	@mkdir $$(UCODE_OUTPUT_DIR)
+  else
+	@mkdir -p $$(UCODE_OUTPUT_DIR)
+  endif
+  # Code file target variables. make does not expand variables within recipes
+  # until the recipe is executed, meaning that all the parts of the recipe would
+  # have the values from the very last microcode in the file. Here, we set
+  # target-specific variables--effectively local variables within the recipe--
+  # to the values from the global variables have right now. We are only
+  # targeting CODE_FILE even though we also want DATA_FILE, because target-
+  # specific variables may not work as expected with multiple targets from one
+  # recipe.
+  $$(CODE_FILE): ARMIPS_CMDLINE:=$$(ARMIPS_CMDLINE)
+  $$(CODE_FILE): CODE_FILE:=$$(CODE_FILE)
+  $$(CODE_FILE): DATA_FILE:=$$(DATA_FILE)
+  # Target recipe
+  $$(CODE_FILE): ./f3dex2.s ./rsp/* ucodes_database.mk $(EXTRA_DEPS) | $$(UCODE_OUTPUT_DIR)
+	@printf "$(INFO)Building microcode: $(NAME): $(DESCRIPTION)$(NO_COL)\n"
+	@$(ARMIPS) -strequ ID_STR "$(ID_STR)" $$(ARMIPS_CMDLINE)
+  ifneq ($(MD5_CODE),)
+	@(printf "$(MD5_CODE) $$(CODE_FILE)" | md5sum --status -c -) && printf "  $(SUCCESS)$(NAME) code matches$(NO_COL)\n" || printf "  $(FAILURE)$(NAME) code differs$(NO_COL)\n"
```

Comment:
An asterisk is needed before `$$(CODE_FILE)` and `$$(DATA_FILE)` in the `printf` for these two lines to tell md5sum that they're binary files. Without it, md5sum incorrectly returns false on my system.